penaltymodel/database.py

```diff
@@ -621,8 +622,18 @@ def isolated_cache(*args, **kwargs):
     This context manager is not reentrant.
 
     """
-    with tempfile.TemporaryDirectory() as d:
-        with threading.RLock():
+    import sys
+
+    if sys.version_info[:2] >= (3, 10):
+        # Added in 3.10
+        # We need ignore_cleanup_errors for Windows, it will still make a "best effort"
+        # to remove the directory.
```

Comment:
Shouldn't simple database file `.close()` be sufficient?

Reply:
I can try it, but it's not clear to me why that would help?

Reply:
Also, I don't think `TemporaryDirectory.close()` exists, unless you mean `TemporaryDirectory.cleanup()`?

Reply:
The reason why `TemporaryDirectory` can't delete the directory on cleanup (on windows) could be a file still opened in that directory. I had [a similar issue](https://github.com/dwavesystems/dwave-cloud-client/blob/5829a538b2045c9fac1bec7fe7888feb6973945d/tests/auth/test_creds.py#L87-L88) in cloud-client.

Reply:
I see. In this case we don't actually have access to the database connection. We could rework the code to make that happen, but it would be a pretty big change.

The connection _should_ be closed when the relevant `PenaltyModelCache` is garbage collected, but either that's not happening or the garbage collection is happening too late. Likely the latter because in the test function that's failing we're using a deprecated interface which returns a function [that accesses the function from the outer namespace](https://github.com/dwavesystems/penaltymodel/blob/5270f593a5f396279476def14f7e30cd743e92a9/penaltymodel/core/interface.py#L165-L167).

Reply:
Ok, I'm not that familiar with how `PenaltyModelCache` is used, but I assumed since it says it's a context manager in docs, the database files will/can be closed on exit.

Reply:
In this case it guarantees that caches instantiated in the context will be isolated. But not necessarily that it will be closed on exit. Though in most cases it is, just one windows test was failing, likely for the above reason.

Reply:
I would expect a context manager to clean up on exit, but that's definitely out of scope for this PR.